### PR TITLE
CSS2D/3DRenderer: Ensure `Object3D.visible` is correctly evaluated in object hierarchies.

### PR DESCRIPTION
**Pull request by lafflan**
…stors are visible

Related issue: #XXXX

**Description**

A clear and concise description of what the problem was and how this pull request solves it.

*This contribution is funded by [Example](https://example.com)*

### CODE REVIEW

**Comment by Mugen87** [comment]
Um, I'm not sure why the CSS renderers behave differently compared to `WebGLRenderer`. Normally, when `Object3D.visible` is set to `false` neither the object nor its descendant nodes should be rendered. 

It's also better to fix this issue in `renderObject()`. Maybe just adding the following line at the top of the method in `CSS3DRenderer`:
```js
if ( object.visible === false ) return;
```
For `CSS2DRenderer` it should be:
```js
if ( object.visible === false || ( _vector.z < - 1 || _vector.z > 1 )) return;
```
The `visible` property only depends from the layers tests like in `WebGLRenderer`.

@yomotsu Do you know why the CSS renderers handle children of invisible objects differently than `WebGLRenderer`?

**Comment by Mugen87** [comment]
For reproduction: https://jsfiddle.net/nhg8durq/

**Comment by Mugen87** [comment]
Um, I'm not sure why the CSS renderers behave differently compared to `WebGLRenderer`. Normally, when `Object3D.visible` is set to `false` neither the object nor its descendant nodes should be rendered. 

It's also better to fix this issue in `renderObject()`. Maybe just adding the following line at the top of the method in `CSS3DRenderer`:
```js
if ( object.visible === false ) return;
```
For `CSS2DRenderer` it should be:
```js
if ( object.visible === false || ( _vector.z < - 1 || _vector.z > 1 )) return;
```
The `visible` property only depends from the layers tests like in `WebGLRenderer`.

@yomotsu Do you know why the CSS renderers handle children of invisible objects differently than `WebGLRenderer`?

**Comment by yomotsu** [comment]
Actually, I'm not entirely sure, but it seems that the potential problem might have gone unnoticed.
The issue is attributed to the following reasons:

- Setting `visible = false` will set `display: none` on the CSS3DObject internally, and its **DOM children** will not be rendered either.
  see: https://github.com/mrdoob/three.js/blob/79c34909e6af9250c38182665e3bd4896149f0ea/examples/jsm/renderers/CSS3DRenderer.js#L259
- However, nested CSS3DObjects are not actually nested in the DOM scene graph but are flattened.
- Therefore, setting `visible = false` on a parent CSS3DObject does not affect its children.

I thought nested CSS3DObjects are also nested in the DOM scene graph and parents `display: none` should affect its children, but not.

Given these points, I think each CSS3DObject needs to check its ancestors' `visible` as this PR does.
Alternatively, CSS3DObjects could have a `.visible` setter that changes the CSS display property ('none' or 'block') of the element and its descendants (no need to check ancestors at every render).

**Comment by yomotsu** [comment]
That makes sense and I tested adding that.

```
function renderObject( object, scene, camera, cameraCSSMatrix ) {

	if ( object.visible === false ) return;	
```

The above will stop the rendering process of children, However, CSS3DObjects will remain rendered without matrix updates because DOM elements are rendered because HTML is rendered in "retained mode", unlike WebGL or Canvas2D.
if we take the above approach, I think all CSS3DObjects must be hidden before rendering, then, show it if `visible === true`

**Comment by Mugen87** [comment]
https://rawcdn.githack.com/lafflan/three.js/patch-1/examples/index.html

**Comment by lafflan** [comment]
Maybe there is another solution that to just change the `Object3D` core code:
```
class Object3D{
  constructor(){
    ...
    // this.visible=true;
    this._visible=true;
  }
  set visible(val) {
    function setChildrenVisible(children) {
      for (let i = 0, l = objects.length; i < l; i++) {
        children[i].visible = val;
        setChildrenVisible(children[i].children);
      }
    }
    this._visible = val;
    setChildrenVisible(this.children);
  }

  get visible() {
    return this._visible;
  }
}
```

**Comment by Mugen87** [comment]
Let's not apply any changes to `Object3D` and focus on the renderers instead.

**Comment by Mugen87** [comment]
The problem with `ancestorIsVisible()` is it performs a lot of redundant hierarchy traversal.

Do you mind implementing the PR differently based on https://github.com/mrdoob/three.js/pull/28293#issuecomment-2097841080?

**Comment by lafflan** [comment]
Done. @Mugen87  @yomotsu

File: examples/jsm/renderers/CSS2DRenderer.js [hunk 1]
```diff
@@ -114,8 +115,21 @@ class CSS2DRenderer {
...
+			for ( let i = 0, l = object.children.length; i < l; i++ ) {
+
+				hideObject( object.children[i] );
+
+			}
+		}
+
 		function renderObject( object, scene, camera ) {
 
+			if ( object.visible === false ) return;
```
Comment (Mugen87) [inline examples/jsm/renderers/CSS2DRenderer.js:131]: It seems the ` object.visible === true ) &&` further below can be removed now.

Comment (yomotsu) [reply examples/jsm/renderers/CSS2DRenderer.js:131]: That's right. The object has to be visible === true below, so there's no need to check again.
Thus, line 138 can be
```diff
- const visible = ( object.visible === true ) && ( _vector.z >= - 1 && _vector.z <= 1 ) && ( object.layers.test( camera.layers ) === true );
+ const visible = ( _vector.z >= - 1 && _vector.z <= 1 ) && ( object.layers.test( camera.layers ) === true );
```

Comment (Mugen87) [reply examples/jsm/renderers/CSS2DRenderer.js:131]: Beautiful!

**Comment by lafflan** [comment]
@Mugen87 There is another problem in `CSS2DRenderer.js `'s `zOrder` method:
```
function zOrder(scene) {
  ...
  const distanceA = cache.objects.get(a).distanceToCameraSquared;
  const distanceB = cache.objects.get(b).distanceToCameraSquared;
  ...
}
```
Due to the cache setting condition that objects must be visible (including their ancestors), when an object is initially invisible ,the `cache.objects.get(object)` will return `undefined`, and the above code will report an error:


So it is necessary to check if the cache is empty and set a default value , some code like this:
```
const cacheDataA = cache.objects.get(a);
const distanceA = cacheDataA ? cacheDataA.distanceToCameraSquared : 0;
const cacheDataB = cache.objects.get(b);
const distanceB = cacheDataB ? cacheDataB.distanceToCameraSquared : 0;
```

**Comment by Mugen87** [comment]
Good catch! Would you file another PR with your fix?

**Comment by lafflan** [comment]
Done. New PR is [#28330](https://github.com/mrdoob/three.js/pull/28330) @Mugen87

**Pull request merged**